### ISSUE TITLE
Feature request: GPU ops for tf.unique; tf.where; and tf.dynamic_partition 

### ISSUE THREAD

**Issue by liuyipei** [tensorflow/tensorflow#5965]
This is a feature request.  As far as I know, all three of them currently do not have GPU ops.

It seems that if we can at least get a GPU implementation of `tf.unique` for integers, then the user can make `tf.where` and `tf.dynamic_partition` manually.  For those of us who are trying to build models that want to mess around with indices rather frequently, this would be incredibly helpful.

**Comment by zheng-xq** [comment]
I marked this contribution-welcome. GPU is not particularly fast for those operations. But it is conceivable that it is faster than copying the data from GPU-to-CPU, operate on CPU, and then copy the data back. 

The kernel itself may need some consideration. For tf.unique, if the range is limited, we can first convert it into a bit-mask, and then gather the active location, through either atomic operations, or a multiple-pass algorithm that figure out the size with preprocessing. 

As for TensorFlow itself, it is a bit tricky to deal with variable-size output Tensor on GPU. I think the best way is to mark them as async-op and wait for the result to come back and then set the tensor size. I can elaborate it a bit more if someone is interested in taking this up.

**Comment by AnishShah** [comment]
Hi @zheng-xq, I would like to work on this.

I was trying to read the source code to understand how to add GPU ops for `tf.unique`. 
After reading different files like `argmax_op.cc` and `softmax_op.cc`, I understood that I need to add `Device` and `functor`. `Functor` will have the logic for the GPU op. Please correct me if I'm wrong.

I read `unique_op.cc` and understood the current logic which is a multiple-pass algorithm. I wanted to know if the new `functor` will have the same logic.
Then, I read `where_op.cc` and  found out that there's already a GPU logic for it. So, I think, we don't need any changes in that.

Thanks! :)

**Comment by ekelsen** [comment]
Probably the best general purpose way (arbitrary data types, arbitrary range) to write unique on gpus is to do a stable sort (that also maintains the index) and then extract the first element from each grouping.

thrust::sort and thrust::unique would do this.  Thrust is already a part of cuda.  The radix sort that would be used the vast majority of time would naturally take advantage of input that has a limited range by limiting the number of passes of the sort.

Downside of thrust::sort is that it is out of place and would require additional memory.  I'm not aware of any high performance sort implementations on the GPU that are in-place.

**Comment by liuyipei** [comment]
I feel that that the primary practical concern being addressed is the gpu-cpu-gpu latencies; the actual time/memory cost on the GPU itself is probably the lesser concern.

**Comment by ekelsen** [comment]
I agree that running it on the GPU will likely be faster than the GPU->CPU->GPU pipeline for non-optimal implementations.  But, the additional memory requirement could break code that works today which might upset some people...

**Comment by ekelsen** [comment]
Note that you would also need to do this: https://github.com/thrust/thrust/blob/1.6.0/examples/cuda/custom_temporary_allocation.cu

to get thrust to use the TF allocator for the temporary storage that it will allocate.

**Comment by dantkz** [comment]
Has this been implemented yet?

**Comment by ekelsen** [comment]
tf.where is done.

tf.unique probably cannot be done - it is already registered as a GPU op, but the calculation happens on the CPU - moving this to the GPU would probably slow down some tiny use cases.  It would be best to create a new op in contrib - unique_gpu or something.

tf.dynamic_partition is not implemented yet.  The easiest way to do it would be to use CUB's radix sort, then scan, then gather.  You could also try to implement the calculation of the output indices as one scan if the total number of partitions was not too large with a custom fancy iterator.

**Comment by dantkz** [comment]
@ekelsen I've been looking for GPU implementation of tf.unique, and it seems one could implement non-stable variant with hash tables using tf operations like scatter_nd_update and gather_nd. However, I don't know if those operations run on GPU?

Furthermore, a unique operation that is applied to each item of a minibatch separately complicates things a bit as scatter works with tf.Variable, but map_fn will produce tensor outputs.

**Comment by ekelsen** [comment]
@dantkz I think the best bet for a GPU version of unique would still be to create a separate op in contrib so that it doesn't mess up the current registration on the host.

scatter_nd and gather_nd do have GPU registrations.  You could try to write a unique using these.  However, I think that using the CUB techniques I outlined would probably be give the best performance across a wide range of input data distributions.

**Comment by isaprykin** [comment]
Hey folks, is it the case that only work related to tf.dynamic_partition is left?

**Comment by dantkz** [comment]
@ekelsen 
It seems that the GPU kernel of tf.unique has memory leak. https://github.com/tensorflow/tensorflow/issues/17155

**Comment by kzjeef** [comment]
@ekelsen @zheng-xq  I recently did some work add a GPU kernel for tf.unique op, I did local benchmark of original c++ hash table and GPU version, it's about 75-100x faster. since during training, tf.unique will take most the time of Opmimizer, I think it will help improve the training performance. 

althrough I noticed there is some comments in `unique_op.cc`:

```
// Fake integer GPU kernels so that the use of Unique in optimizers (to
// de-duplicate sparse gradient indices) does not conflict with gradients being
// located on a GPU. These kernels run on the CPU, their inputs and outputs
// residing in host (not GPU) memory.
```

I'm not sure what this comment means, did it means this op cannot implement in GPU ? 
btw, I did this work and test based on Unique op (not the recently added UniqueV2 Op).

**Comment by ekelsen** [comment]
It means that we actually would really like GPU kernels for unique, but have to resort to running CPU implementations because there is no GPU option.  If you have a GPU implementation, then that would be great, please submit a CL.

**Comment by aldazero** [comment]
Hi! any progress here? tf.unique  is very time consuming on CPU :(

**Comment by antant-shenzhen-2013** [comment]
+1

**Comment by fightbob** [comment]
Hi @zheng-xq , I write tf.unique GPU code, but it can't run concurrently. For 55 independent unique tasks, it runs one by one on the GPU. It seems all tasks  on GPU run sequentially. The reason is tensorflow only creates one compute-stream as you mentioned at [stackoverflow](https://stackoverflow.com/a/36108627/5193007). Tensorflow can run GPU op concurrently?

**Comment by fightbob** [comment]
> @fightbob as far as I know, tf.unique always runs on CPU, even if you force-place it on GPU.

@dantkz Yes,  always run on CPU, guessing it's associated with TF Op Register mechanism.  I remove all tf.Unique cpu codes, and only stay gpu code. It's tricky but works. If tf.Unique can run parallelly on mutiple gpu stream, I will take better ways instead of the tricky. Do you know how to call TF op concurrently on multiple stream?

**Comment by liob** [comment]
any news? This is the major performance hit our pipeline takes during training. Approximately 50 % of training time is spent in the unique_with_counts op. That it can be done efficiently has been demonstrated by pytorch.

**Comment by tatianashp** [comment]
@sanjoy Can you comment on this?

**Comment by sanjoy** [comment]
We (TensorFlow team) don't have enough cycles to work on this right now, but I'll happily review a PR.

**Comment by tilakrayal** [comment]
@liuyipei,
The GPU kernel for dynamic partition was introduced in TensorFlow 1.5 and as per the [PR](https://github.com/tensorflow/tensorflow/pull/13905) raised it was fixed by 1.6.0's final release which was upgraded to the CUB library version.

The issue is the tensor in embeds and mol_idx do not always have the same first dimension and whether or not the one with the wrong dimension size is from embeds or mol_idx (or both) for sure since these sizes will vary based on the batch, but we  suspect it is from mol_idx based on typical sizes when the error does not occur. One possible cause may be the difference in dtypes. embeds contains tensors of either float32 or float64, and mol_idx contains tensors of int32.
